File: common-core/src/main/java/io/quartic/common/ApplicationDetails.java
package io.quartic.common;

import org.immutables.value.Value;

@SweetStyle
@Value.Immutable
public abstract class ApplicationDetails {
    protected abstract Class<?> clazz();

    @Value.Lazy
    public String name() {
        final String name = clazz().getSimpleName();
        return name.replaceAll("Application$", "");
    }

    @Value.Lazy
    public String version() {
        final String version = clazz().getPackage().getImplementationVersion();
        return ((version == null) ? "unknown" : version);
    }

    @Value.Lazy
    public String javaVersion() {
        return System.getProperty("java.version");
    }
}

File: common-core/src/main/kotlin/io/quartic/common/ApplicationDetails.kt
package io.quartic.common

class ApplicationDetails(private val clazz: Class<*>) {
    val name: String
        get() = clazz.simpleName.replace("Application$".toRegex(), "")

    val version: String
        get() = clazz.`package`.implementationVersion ?: "unknown"

    val javaVersion: String
        get() = System.getProperty("java.version")
}
